raster/src/main/scala/geotrellis/raster/io/geotiff/OverviewStrategy.scala

```diff
@@ -16,20 +16,97 @@
 
 package geotrellis.raster.io.geotiff
 
+import geotrellis.raster.CellSize
+
+import scala.collection.Searching._
+
+/**
+  * Options ported from GDAL (https://gdal.org/programs/gdalwarp.html#cmdoption-gdalwarp-ovr)
+  *  for automatic selection from among available overviews
+  */
 sealed trait OverviewStrategy
+
+object OverviewStrategy {
+  def DEFAULT = AutoHigherResolution
+
+  /**
+    * Select appropriate overview given the strategy.
+    *
+    * WARN: this function assumes that CellSizes are sorted. It interprets idx 0 as the position with the highest CellSize.
+    *
+    * Unless a particular strategy suggests otherwise, this method will clamp the returned
+    * index to the range of overviewCS.
+    * @param overviewCS
+    * @param desiredCS
+    * @param strategy
+    * @return
+    */
+  def selectOverview(
+    overviewCS: List[CellSize],
+    desiredCS: CellSize,
+    strategy: OverviewStrategy
+  ): Int = {
+    val maybeIndex = strategy match {
+      case Level(overviewIdx) =>
+        overviewIdx
+      case Auto(n) =>
+        selectIndexByProximity(overviewCS, desiredCS, 0.5) + n
+      case Base =>
+        overviewCS.indexOf(overviewCS.min)
+      case AutoHigherResolution =>
+        val idx = selectIndexByProximity(overviewCS, desiredCS, 1.0)
+        // AutoHigherResolution defaults to the closest level
+        // idx > overviewCS.size means, that idx = overviewCS.size (last idx + 1)
+        if (idx > overviewCS.size) idx - 1 else idx
```

Comment:
We need to know where is the highest resolution to interpret the index the `selectIndexByProximity` function. If the input list is sorted, `idx > overviewCS.size` means the lower resolution. In case it is reversed, it means the higher resolution. If the list is not sorted, this index makes no sense for its interpretation.